hareHook: init

onemoresuza opened the pull request:
## Description of changes

The `hareHook` package does the following:

1. Sets `HARECACHE` to `$(mktemp -d)`;
2. Allows for cross-compilation out of the box, by wrapping the hare binary in a script that passes the `-a` flag for it;
3. Provides the `hareBuildType` attribute, which accepts either `release` (default) or `debug` for passing the `-R` for the wrapper script;
4. Removes the pollution of logs by passing the `-q` flag to the wrapper script;
5. Provides the `hare-native` script for when there's a need to run compiled binaries on the build process; and
6. Appends Hare's stdlib to `HAREPATH` only after all third party libraries are added, so that it may be extended or shadowed by them.

### TODO

- [x] Fix cross-compilation (Fixed, thanks to @uninsane in https://github.com/onemoresuza/nixpkgs/pull/1#issue-2182138217.)
	The hook does not seem to pass the `buildPlatform` `hare` binary when
	cross-compiling, but the one from `hostPlatform`.
- [x] Add an entry to the nixpkgs manual.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/hare-build-framework-decision-looking-for-feedback/41189/1

uninsane commented on pkgs/by-name/ha/hare/package.nix:
```diff
@@ -146,9 +146,8 @@ stdenv.mkDerivation (finalAttrs: {
-  setupHook = ./setup-hook.sh;
-
   passthru = {
+    hook = callPackage ./hook.nix { hare = finalAttrs.finalPackage; };
```
i expect passthru attributes don't get spliced, which could explain why cross compiling doesn't work (`hare.hook` would represent the derivation built for the host platform, regardless of which `nativeBuildInputs`/`buildInputs` set you reference it from).

i'm not sure the proper fix. possibly some amount of `lib.mkScope`, `lib.recurseIntoAttrs`, or making the `hare` package itself function as the setup hook (with some `dontHareConfigure` option to disable it). i'll poke around at how meson does this.

uninsane replied:
@onemoresuza i think i fixed this to support cross compilation (i.e. `pkgsCross.aarch64-multiplatform.<pname>`), also in the process made it so that `hare` _is_ the setupHook (much like how `meson`, `cmake` work). tested it with `bonsai`, and also built but didn't run `haredo`.

https://github.com/onemoresuza/nixpkgs/compare/hare-hook...uninsane:nixpkgs:hare-hook-cross

uninsane commented on pkgs/by-name/ha/hare/setup-hook.sh:
```diff
@@ -1,9 +1,26 @@
+    HAREFLAGS="${HAREFLAGS-} ${hareDefaultFlags[*]}"
+    export HAREFLAGS
+    echoCmd 'HAREFLAGS' "$HAREFLAGS"
+    makeFlagsArray+=(
+        HARECACHE="$HARECACHE"
+        HAREFLAGS="$HAREFLAGS"
+    )
+    echoCmd 'makeFlagsArray' "${makeFlagsArray[@]}"
```
it's unfortunate that plumbing `HAREFLAGS` through to the actual `hare` invocation requires cooperation from every layer (correct me if i'm wrong): not only do we have to call `make HAREFLAGS="$HAREFLAGS"`, but also the project's Makefile needs to be aware of that and invoke hare with `hare $(HAREFLAGS)`. then as more hare projects come along using things other than make (like haredo), we have to make sure each new build tool is properly integrated...

i wonder if we should actually just wrap the `hare` executable so that it appends `$HAREFLAGS` (or even our own `NIX_HAREFLAGS`) directly, and then just let each package/builder do its own thing rather than try to plumb things through each layer cooperatively?

onemoresuza replied:
I like your idea of wrapping the hare executable: We avoid the need for integration with each new build tool, while achieving the same, that is, passing HAREFLAGS to the executable.

I also think that we can split what's being done in the setup-hook:

1. `HARECACHE` being only an environment variable still allows the hare binary to use, while also allowing build tools to have access to its value;
2. The *default* `HAREFLAGS` now are `-q`, for not polluting the build logs; `-R`, for building on release mode; and `-a<arch>`, for cross-compiling. I believe we should add a `hareDontUseReleaseMode` --- or something equivalent --- attribute, defaulting to `false`, to allow derivation to disable the `-R`, since, although rare, it may rely on the `debug` module.

onemoresuza replied:
I've tried a few approaches regarding wrapping the hare binary to no avail:

1. Creating a `hare() {}` bash function: It works on packages the call directly from the shell environment --- *e. g.*, [haredo](https://github.com/NixOS/nixpkgs/blob/85d3f97277c0a0b5b542951e8990b4bd2bbe2a26/pkgs/by-name/ha/haredoc/package.nix#L30) and [haredoc](https://github.com/NixOS/nixpkgs/blob/85d3f97277c0a0b5b542951e8990b4bd2bbe2a26/pkgs/by-name/ha/haredoc/package.nix#L30) ---, but not for packages that call it from a build tool --- *e. g.*, [bonsai](https://github.com/NixOS/nixpkgs/blob/85d3f97277c0a0b5b542951e8990b4bd2bbe2a26/pkgs/by-name/bo/bonsai/package.nix) with `make` ---, so it's a no go.
2. Creating `writeShellApplication` with the hare binary wrapped in it and adding it to `PATH` through the setup-hook the same way `HAREPATH` is set: Although the script ends up in `PATH`, it ends up after hare's derivation path, thus the latter has precedence over the former, which ends up not being used.

Any ideas on this, @uninsane? Perhaps we could try again a `hare.hook` attribute passed through `passthru`, since it would allow us to not have the unwrapped hare binary `PATH`, but we would still need to solve the cross-compilation brought upon the first iteration.

uninsane replied:
my first thought would be to use `makeWrapper` when building hare itself:
```nix
# hare/package.nix:
{ ... }:
stdenv.mkDerivation {
  pname = "hare";
  ...
  nativeBuildInputs = [ makeWrapper ... ];
  postInstall = ''
    wrapProgram $out/bin/hare \
      --add-flags '$NIX_HAREFLAGS'  # quoted so it won't expand at the time we make the program, but hopefully at the time the wrapper is invoked
  '';
};
```

like, don't even create a new derivation for this: just add that `postInstall` to the existing `hare` derivation. if that doesn't fit the bill i would have to look at our `cc` wrapper for inspiration.

onemoresuza replied:
Neither `--add-flags` nor `--append-flags` would work because the added `$NIX_HAREFLAGS` would be, respectively, before the subcommand --- *e. g.*, `hare $NIX_HAREFLAGS <cmd>` ---, causing a `unrecognized option: ` error, or after all arguments --- *e. g.*, `hare <cmd> $@ $NIX_HAREFLAGS` ---, which would make it be treated as an positional argument any time a path was passed and we cannot guarantee that every package invocation of the hare binary would be called without passing a path to it --- like so: `hare <cmd> <flags>`.

If there is a way to select at which index the appended flag would go, we would be able to use your `wrapProgram` solution.

onemoresuza replied:
I've tried creating a different derivation and adding it to `pkgs/top-level/all-packages.nix` as `hareHook` and it works with both native and cross-compilation; it, however, when set as an attribute of hare's `passthru` has the same issue with cross-compilation that the `hare.hook` solution had. Moreover, if we were to change to a different derivation, the current `setupHook` of hare would had to go to it and it would break hare's functionality on a `nix-shell`, since:

1. For `HAREPATH` to be set, the user would had to add `hareHook` to the nix-shell;
2. We would also need to find a way to disabling passing `-qRa<arch>` --- perhaps a pattern like `noDefaultFlags` ---, since, in a nix-shell, the user would prefer fine control of which options are passed to the binary.

uninsane replied:
as for the `noDefaultFlags` stuff, i notice several build hooks support attributes like `mesonBuildType = "plain"|"debug"|"release"|...` (e.g. `cmakeBuildType` for cmake, `cargoBuildType` for rust, `dubBuildType` for Dlang),  so `hareBuildType = "plain"|"release"` may be an idiomatic way to control whether we pass the `-R` flag or not.

hopefully i'll have some time to dig into the wrapper stuff this weekend.

onemoresuza commented on pkgs/by-name/ha/haredo/package.nix:
```diff
@@ -26,22 +26,19 @@ stdenv.mkDerivation (finalAttrs: {
     scdoc
   ];
 
+  env.PREFIX = builtins.placeholder "out";
+
```
It is. After refactoring it for cross-compilation, the boostrap script --- which uses the `PREFIX` env var --- is no longer used.

onemoresuza replied:
I've rebased it and removed the attribute.

onemoresuza committed:
hareHook: init

Co-authored-by: Colin <colin@uninsane.org>

onemoresuza committed:
treecat: format with nixfmt-rfc-style

onemoresuza committed:
treecat: make use of hareHook

onemoresuza committed:
haredoc: format with nixfmt-rfc-style

onemoresuza committed:
haredoc: make use of hareHook

onemoresuza committed:
bonsai: format with nixfmt-rfc-style

onemoresuza committed:
bonsai: make use of hareHook

onemoresuza committed:
hareThirdParty.hare-png: format with nixfmt-rfc-style

onemoresuza committed:
hareThirdParty.hare-compress: format with nixfmt-rfc-style

onemoresuza committed:
hareThirdParty.hare-compress: make use of hareHook

onemoresuza committed:
hareThirdParty.hare-png: make use of hareHook

onemoresuza commented:
@uninsane, @starzation, I've been testing a `hareHook` package on [this fork](https://github.com/onemoresuza/nixpkgs/tree/hare-setup-hook).

It uses @uninsane's suggestion of having a hare wrapper, which passes the value of `NIX_HAREFLAGS` to the hare binary, while also allowing passing any other arguments to it. Moreover, as also suggested by @uninsane, it implements a `hareBuildType`, which can either be `release` (the default when not set or empty) or `debug` (which compiles the binary with the debug module enabled).

I've also patched `haredoc`, `treecat`, `bonsai`, `hare-compress` and `hare-png` to make use of it and tested both its native (x86_64) and cross compilation (aarch64 and riscv64) and they both work. I've chosen those packages since they seem to be representative of the current Hare ecosystem on nixpkgs:

1. `haredoc` is built using the `hare` command directly;
2. `treecat` uses `haredo`;
3. `bonsai` uses `make`;
4. `hare-compress` is a library; and
5. `hare-png` is a library with propagated build inputs.

If you're okay with its implementation, I could drop the current commits of this PR and bring the ones from the aforementioned fork to it.

What do you think?

onemoresuza committed:
hareThirdParty.hare-json: format with nixfmt-rfc-style

onemoresuza committed:
hareThirdParty.hare-json: make use of hareHook

onemoresuza committed:
hareThirdParty.hare-toml: format with nixfmt-rfc-style

onemoresuza committed:
hareThirdParty.hare-toml: make use of hareHook

onemoresuza committed:
hareThirdParty.hare-ev: format with nixfmt-rfc-style

onemoresuza committed:
hareThirdParty.hare-ev: make use of hareHook

onemoresuza committed:
hareThirdParty.hare-ssh: format with nixfmt-rfc-style

onemoresuza committed:
hareThirdParty.hare-ssh: make use of hareHook

onemoresuza committed:
haredo: make use of hareHook

onemoresuza committed:
himitsu: make use of hareHook

onemoresuza commented:
Since it has been almost a week without an answer, I'll be using the
`hareHookPkg` branch for this PR.

For that  I'll:

1. Drop the commits on this PR; and
2. `git am -3` the formatted patches of the aforementioned patch.

@uninsane, since I'll be making use of code and ideas proposed by you on those
patches, I've put you as a co-author of the `hareHook: init` commit. If you wish
to not be a co-author or to either change the name or email used ("Colin" and
"colin at uninsane dot org", respectively), please, let me know.

Moreover, the old changes are not lost: I've saved them as a gist in the format
produced by `git format-patch`. The patches may be found
[here](https://gist.github.com/onemoresuza/eb4cdcb3a5ba36743f4af42035e1d161).

onemoresuza commented:
Result of `nixpkgs-review pr 295155` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>himitsu-firefox</li>
  </ul>
</details>
<details>
  <summary>17 packages built:</summary>
  <ul>
    <li>bonsai</li>
    <li>hare</li>
    <li>hare.man</li>
    <li>hareHook</li>
    <li>hareThirdParty.hare-compress</li>
    <li>hareThirdParty.hare-ev</li>
    <li>hareThirdParty.hare-json</li>
    <li>hareThirdParty.hare-png</li>
    <li>hareThirdParty.hare-ssh</li>
    <li>hareThirdParty.hare-toml</li>
    <li>haredo</li>
    <li>haredo.man</li>
    <li>haredoc</li>
    <li>haredoc.man</li>
    <li>himitsu</li>
    <li>treecat</li>
    <li>treecat.man</li>
  </ul>
</details>

uninsane commented on pkgs/by-name/ha/hare/hook.nix:
```diff
@@ -0,0 +1,62 @@
+}:
+let
+  arch = stdenv.targetPlatform.uname.processor;
+  harePropagationInputs = builtins.filter (
+    drv:
+    builtins.any (inputName: lib.getName drv == inputName) [
+      "harec"
+      "qbe"
+    ]
+  ) hare.nativeBuildInputs;
```
it may be less fragile to expose these in hare's `passthru` attribute and then reference them here as `harePropagationInputs = [ hare.passthru.harec hare.passthru.qbe ];`

uninsane commented on pkgs/by-name/ha/hare/setup-hook.sh:
```diff
@@ -1,9 +1,36 @@
+echoCmd "hare" "$(command -v hare)"
+echoCmd "hare-native" "$(command -v hare-native)"
+'
+prePhases="hareSetStdlibPhase hareInfoPhase"
```
i think this should be `prePhases+= ...`, i.e. append instead of assign.

uninsane reviewed (commented):
ah, i'm really sorry for not following through with this. this seems like a fine approach though 👍

onemoresuza committed:
harec: format with nixfmt-rfc-style

onemoresuza committed:
harec: expose qbe through passthru

harec and hare should always have the same qbe package. By exposing the
latter under harec's passthru attribute and using it as hare's qbe, we
assure that they will always hare the same qbe package even if there's a
need for a local overlay.

onemoresuza commented:
>
> ah, i'm really sorry for not following through with this. this seems like a fine approach though 👍
>

No worries, this is voluntary work after all.

> > +  harePropagationInputs = builtins.filter (
> +    drv:
> +    builtins.any (inputName: lib.getName drv == inputName) [
> +      "harec"
> +      "qbe"
> +    ]
> +  ) hare.nativeBuildInputs;
>
> it may be less fragile to expose these in hare's `passthru` attribute and then reference them here as `harePropagationInputs = [ hare.passthru.harec hare.passthru.qbe ];`

You're right: I've changed `harePropagationInputs` to use the
`passthru`'s attributes of hare (`harec` and `qbe`), being the latter a
passthru attribute from harec, since it and hare must always have the
same qbe.

>
> >  }
>  
> +# Hare's stdlib should come after its third party libs, since the latter may
> +# expand or shadow the former.
> +readonly hareSetStdlibPhase='
> +addToSearchPath HAREPATH ***@***.***_stdlib@"
> +'
> +readonly hareInfoPhase='
> +echoCmd "HARECACHE" "$HARECACHE"
> +echoCmd "HAREPATH" "$HAREPATH"
> +echoCmd "hare" "$(command -v hare)"
> +echoCmd "hare-native" "$(command -v hare-native)"
> +'
> +prePhases="hareSetStdlibPhase hareInfoPhase"
>
> i think this should be `prePhases+= ...`, i.e. append instead of assign.
>

Changed to `prePhases+=("hareSetStdlibPhase" "hareInfoPhase")`

onemoresuza committed:
doc/hare: init

onemoresuza commented:
I've added the backport label to lessen the burden on maintainers: If this is added to unstable, but not to 24.05, a maintainer would have to maintain a package with the old framework and with the new one.

pbsds commented on pkgs/by-name/ha/hare/hook.nix:
```diff
@@ -0,0 +1,56 @@
+  arch = stdenv.targetPlatform.uname.processor;
+  harePropagationInputs = builtins.attrValues { inherit (hare) harec qbe; };
+  hareWrappedScript = writeShellApplication {
+    # `name` MUST be `hare`, since its role its to replace the hare binary.
```
```suggestion
    # `name` MUST be `hare`, since its role is to replace the hare binary.
```

onemoresuza replied:
Fixed the typo.

pbsds reviewed (commented):
Result of `nixpkgs-review pr 295155` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>himitsu-firefox</li>
  </ul>
</details>
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>17 packages built:</summary>
  <ul>
    <li>bonsai</li>
    <li>hare</li>
    <li>hare.man</li>
    <li>hareHook</li>
    <li>hareThirdParty.hare-compress</li>
    <li>hareThirdParty.hare-ev</li>
    <li>hareThirdParty.hare-json</li>
    <li>hareThirdParty.hare-png</li>
    <li>hareThirdParty.hare-ssh</li>
    <li>hareThirdParty.hare-toml</li>
    <li>haredo</li>
    <li>haredo.man</li>
    <li>haredoc</li>
    <li>haredoc.man</li>
    <li>himitsu</li>
    <li>treecat</li>
    <li>treecat.man</li>
  </ul>
</details>

also tested on x86_64-linux:
```
nom-build nixpkgs -A pkgsCross.riscv64.bonsai -A pkgsCross.riscv64.hareHook -A pkgsCross.riscv64.hareThirdParty.hare-compress -A pkgsCross.riscv64.hareThirdParty.hare-ev -A pkgsCross.riscv64.hareThirdParty.hare-json -A pkgsCross.riscv64.hareThirdParty.hare-png -A pkgsCross.riscv64.hareThirdParty.hare-ssh -A pkgsCross.riscv64.hareThirdParty.hare-toml -A pkgsCross.riscv64.haredo -A pkgsCross.riscv64.haredoc -A pkgsCross.riscv64.himitsu -A pkgsCross.riscv64.treecat -A pkgsCross.aarch64-multiplatform.bonsai -A pkgsCross.aarch64-multiplatform.hareHook -A pkgsCross.aarch64-multiplatform.hareThirdParty.hare-compress -A pkgsCross.aarch64-multiplatform.hareThirdParty.hare-ev -A pkgsCross.aarch64-multiplatform.hareThirdParty.hare-json -A pkgsCross.aarch64-multiplatform.hareThirdParty.hare-png -A pkgsCross.aarch64-multiplatform.hareThirdParty.hare-ssh -A pkgsCross.aarch64-multiplatform.hareThirdParty.hare-toml -A pkgsCross.aarch64-multiplatform.haredo -A pkgsCross.aarch64-multiplatform.haredoc -A pkgsCross.aarch64-multiplatform.himitsu -A pkgsCross.aarch64-multiplatform.treecat
```

notes:

* 86ae868f0340a571696bb3f9dd6b2fa5153261d2 seems to unset HARECACHE without adding hareHook, is this intentional?

pbsds commented:
I'm in favor of backporting this, but i will ask the release managers before doing so.

Just a quick idea which may avoid that:
The only breaking change as i see it is the _switch_ to `hareHook` where `hare` is wrapped with `-a $arch`. Is is it possible to instead only wrap `hare` in `pkgsCross` only, and leave `hare` unwrapped in `pkgs`? Then there would be no breaking change and we could backport it even after release

onemoresuza committed:
himitsu: format with nixfmt-rfc-style

onemoresuza commented:
> * [86ae868](https://github.com/NixOS/nixpkgs/commit/86ae868f0340a571696bb3f9dd6b2fa5153261d2) seems to unset HARECACHE without adding hareHook, is this intentional?

That was a mistake: I've rebased and moved the removal of the `preBuild` phase to when `hareHook` is added to `nativeBuildInputs`.

onemoresuza commented on doc/languages-frameworks/hare.section.md:
Added to index.md.

onemoresuza commented:
> I'm in favor of backporting this, but i will ask the release managers before doing so.
> 
> Just a quick idea which may avoid that: The only breaking change as i see it is the _switch_ to `hareHook` where `hare` is wrapped with `-a $arch`. Is is it possible to instead only wrap `hare` in `pkgsCross` only, and leave `hare` unwrapped in `pkgs`? Then there would be no breaking change and we could backport it even after release

Is there a way to distinguish when the derivation is in `pkgsCross`? If there is, I think there would no trouble in only wrapping the hare binary if in it.

pbsds commented:
`stdenv.buildPlatform == stdenv.hostPlatform` may do the trick

onemoresuza commented:
@pbsds, I've managed to do the distinction by using `stdenv.buildPlatform == stdenv.targetPlatform` on `pkgs/by-name/ha/hare/hook.nix` --- there were some `binutils` errors when using `hostPlatform`. However, even on native builds, the hare binary would need to be wrapped for the `-q` and `-R` flags --- quieter logs and release mode respectively.

I think is better to wait for the answer of the backporting managers, since the hare binary would have to be wrapped either way.

pbsds reviewed (approved):
LGTM

@nixos/nixos-release-managers #303285: 
we want to backport this to 24.05 before release. This is technically a breaking change since `hareHook` is now to be used instead of `hare` when building hare packages, but the hare ecosystem is still quite small, and backporting will massively reduce maintenance burden. I defer the backport decision to you, can you give us the go-ahead? 

The non-breaking change alternative is to wrap `${hare}/bin/hare`, which this PR tries to avoid

_(since i find `@`ing teams to rarely work: @wegank @Mic92)_

pbsds commented:
Touching the 24.05 release notes after release is sadly a no-go, please move the note to 24.11.

To lessen the maintenance burden we could merge this, then backport `hareHook` while not backporting the removal of the setup hook in `hare`, making the both ways still work on 24.05.

onemoresuza commented:
I've moved the release notes to 24.11. Should the backport label be removed since it wont be an automatic backport?

pbsds commented:
Thank you for all your hard work!
